text/0063-precreated-roles.md

```diff
@@ -0,0 +1,182 @@
+
+This RFC proposes a mechanism by which roles creation is replaced with an offline reporting mechanism. Application developers write their application as usual, using standard constructs and standard `grant` calls. Then, when this feature is switched on, `Role` resources are no longer synthesized. Instead, a report is generated containing all Roles and policies that *would* have been created, and synthesis fails. This report can be given to the central IT organization, who can create and configure the roles as reported. Afterwards, the app developer plugs in the names of the roles that the IT organization created, and synthesis succeeds (assuming all would-be Roles have name assigned).
+
+## README: prevent creation of IAM Roles
+
+In normal operation, L2 constructs in the AWS Construct Library will automatically create IAM Roles for resources that require them (like Execution Roles for AWS Lambda Functions and Instance Roles for EC2 instances), and assign least-privilege permissions based on the `grant`s and integrations you define.
+
+If you work in an environment that does not allow definition of IAM Roles by application developers, you can disable
+this behavior by calling `iam.Role.preventRoleCreation()` on the scope at which you want to prevent roles from being
+created, before defining the infrastructure of your application. Example:
+
+```ts
+import { App, aws_iam as iam } from 'aws-cdk-lib';
+import { MyStack } from './my-stack';
+
+const app = new App();
+
+// Disable synthesis of IAM Roles in the entire app
+iam.Role.preventRoleCreation(app);
+
+new MyStack(app, 'MyStack', {
+  // ...
+});
+app.synth();
+```
+
+The next time you run `cdk synth`, a file named `cdk.out/iam-roles.txt` will be created, containing a report of all roles that the CDK app would have created, and the permissions that would be added to them (based on `grant` methods).
+
+For example, `iam-roles.txt` might look like this:
+
+```text
+<missing role> (/MyStack/MyLambda/Role)
+    {
+        "Effect": "Allow",
+        "Action": ["s3:GetObject"],
+        "Resource": ["arn:PARTITION:s3:REGION:ACCOUNT:${/MyStack/MyBucket}/*",
+    }
+```
+
+If there are any roles marked as `<missing role>` (which means there is no well-known role name associated with them yet), synthesis will fail.
+
+Give this report to your IT operators, and ask them to create roles with the required permissions (this will
+require specifying wildcards for names of resources that haven't been created yet).
+
+When your IT department comes to you, they will have created a role with a well-known name. For example, `LambdaRole`.
+Plug that name into your `preventRoleCreation`:
+
+```ts
+iam.Role.preventRoleCreation(app, {
```

Comment:
I'm not sure I follow. It sure looks like no Roles are ever created, and that's exactly what we're trying to prevent, so the name seems apt to me.

Reply:
I kind of think the name should be different as well. I kind of see this as two separate features.

1. Create a report of all roles created and their permissions
2. Prevent role creation and provide an easier way to import the precreated roles.

For example I can see a use case for users wanting to generate the report, but don't want to prevent the role creation. In that case it would be something like:

```ts
iam.Role.preventRoleCreation(app, {
  preventRoleCreation: false,
  // or
  // generateReportOnly: true,
});
```

In which case the name doesn't really make sense.

Reply:
@rix0rrr it makes sense in the case where we are generating the report, yes. But the naming makes less sense when developers are supplying well-known names to the construct, i.e.:

```ts
iam.Role.preventRoleCreation(app, {
  '/MyStack/MyLambda/Role': 'LambdaRole',
});
```

we're no longer just "preventing" the role creation. We're also overriding it with a pre-created role.

I'm not super against it. But I think `override` makes more sense.